workflow-rx2/src/main/java/com/squareup/workflow/rx2/DiscreteStateWorkflow.kt

```diff
@@ -0,0 +1,249 @@
+ *    ) {
+ *      override fun launch(
+ *        initialState: MyState,
+ *        workflows: WorkflowPool
+ *      ) : Workflow<MyState, MyEvent, MyResult> {
+ *        workflows.register(nestedReactor)
+ *        return workflows.discreetStateWorkflow(initialState, workflows, ::nextState)
+ *      }
+ *
+ * and in your `nextState` method, use
+ * [onNextDelegateReaction][EventSelectBuilder.onWorkflowUpdate] to wait for the nested
+ * workflow to do its job:
+ *
+ *    is Delegating -> events.select {
+ *      workflows.onNextDelegateReaction(state) {
+ *        when (it) {
+ *          is EnterState -> EnterState(state.copy(delegateState = it.state))
+ *          is FinishWith -> when (it.result) {
+ *            is DoSomething -> EnterState(DoingSomething)
+ *            is DoSomethingElse -> EnterState(DoingSomethingElse)
+ *          }
+ *        }
+ *      }
+ *    }
+ *
+ * If you need to handle other events while the workflow is running, remember to call
+ * [WorkflowPool.abandonWorkflow] if you leave while the nested workflow is still running!
+ *
+ *    is Delegating -> events.select {
+ *      workflows.onNextDelegateReaction(state) {
+ *        when (it) {
+ *          is EnterState -> EnterState(state.copy(delegateState = it.state))
+ *          if FinishWith -> when (it.result) {
+ *            is DoSomething -> EnterState(DoingSomething)
+ *            is DoSomethingElse -> EnterState(DoingSomethingElse)
+ *          }
+ *        }
+ *      }
+ *
+ *      onEvent<Cancel> {
+ *        workflows.abandonDelegate(state.id)
+ *        EnterState(NeverMind)
+ *      }
+ *    }
+ *
+ * To accept events for nested workflows, e.g. to drive a UI, define
+ * [com.squareup.workflow.Renderer]s for both `S` and each of its
+ * [Delegating][com.squareup.workflow.Delegating] subtypes. [WorkflowPool.input] can be used by
+ * renderers to route events to any running workflow.
+ */
+typealias StateTransitionFunction<S, E, O> =
+    (state: S, events: EventChannel<E>, workflows: WorkflowPool) -> Single<out Reaction<S, O>>
+
+/**
+ * Implements a [Workflow] using a [StateTransitionFunction].
+ * Use this to implement [WorkflowPool.Launcher.launch].
+ *
+ * The react loop runs inside a [workflow coroutine][workflow].
+ *
+ * The [initial state][initialState], and then each [subsequent state][EnterState], are all sent
+ * to the builder's state observable.
+ *
+ * _Note:_
+ * If the state transition function immediately returns [FinishWith], the last state may not be
+ * emitted. See [com.squareup.workflow.discreteStateWorkflow].
+ *
+ * ## Naming
+ *
+ * To help with debugging, this function accepts a [name]. This name will
+ * be used in exceptions thrown from the workflow. Additionally, if coroutine debugging is enabled,
+ * the name will show up in stack traces and thread names.
+ */
+fun <S : Any, E : Any, O : Any> WorkflowPool.discreteStateWorkflow(
+  initialState: S,
+  name: String? = null,
+  nextState: StateTransitionFunction<S, E, O>
```

Comment:
nit: this parameter is not a state, so `nextState` is a misleading name.
Would also reorder these so the optional param is last.

Reply:
The function parameter must come last so it can take a block, that's a kotlin syntax limitation.